### PR TITLE
build: add AWS OFI plugin for NCCL GPU based OFI/libfabric builds 

### PR DESCRIPTION
**Pull request by karlonw**
## Description

This PR updates the Dockerfile-base-gpu build target, when using NCCL and OFI, to also build the AWS OFI plugin, which allows NCCL to utilize libfabric as the provider, rather than falling back to TCP when run on Slingshot 11(SS11) systems, which are not IB verbs based. This PR also provides a helper script to the entrypoint, such that if `/host` is bind-mounted into the container, the helper script will attempt to locate the host's libfabric libraries, and prepend them to the `LD_LIBRARY_PATH`  in order to use the most appropriate libfabric for the system.

## Checklist

- [ ] Bump VERSION to make the pushed images are tagged with the right version.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.

## Testing
```
karlon@nid000001:~/scr> env NCCL_DEBUG=INFO CUDA_VISIBLE_DEVICES=0,1,2,3 srun --mpi=pmi2 -n 8 --ntasks-per-node=4 -p grizzly --gpu-bind=closest singularity run --nv --bind /scratch2,/usr:/host/usr /scratch2/karlon/new/detAI-cuda-11.3-pytorch-1.10-tf-2.8-gpu-nccl-0.19.4.sif python /scratch2/karlon/tf-hvd/tensorflow2_mnist.py 2>&1 |egrep -i 'provider|#12.0|training'
nid000019:51404:52019 [0] NCCL INFO NET/OFI Selected Provider is cxi
nid000019:51338:52013 [2] NCCL INFO NET/OFI Selected Provider is cxi
nid000019:51339:52015 [1] NCCL INFO NET/OFI Selected Provider is cxi
nid000019:51340:52016 [3] NCCL INFO NET/OFI Selected Provider is cxi
nid000020:70404:71079 [2] NCCL INFO NET/OFI Selected Provider is cxi
nid000020:70473:71081 [0] NCCL INFO NET/OFI Selected Provider is cxi
nid000020:70403:71082 [1] NCCL INFO NET/OFI Selected Provider is cxi
nid000020:70405:71083 [3] NCCL INFO NET/OFI Selected Provider is cxi
Step #1200      Loss: 0.009821
Step #1200      Loss: 0.049917
Step #1210      Loss: 0.035170
Step #1210      Loss: 0.027505
Step #1220      Loss: 0.055104
Step #1220      Loss: 0.004597
Step #1230      Loss: 0.006290
Step #1230      Loss: 0.029581
Step #1240      Loss: 0.012338
Step #1240      Loss: 0.045299
time to complete training: 17.1268 seconds
karlon@nid000001:~/scr> env NCCL_DEBUG=INFO CUDA_VISIBLE_DEVICES=0,1,2,3 srun --mpi=pmi2 -n 8 --ntasks-per-node=4 -p grizzly --gpu-bind=closest singularity run --nv --bind /scratch2 /scratch2/k
arlon/new/detAI-cuda-11.3-pytorch-1.10-tf-2.8-gpu-nccl-0.19.4.sif python /scratch2/karlon/tf-hvd/tensorflow2_mnist.py 2>&1 |egrep -i 'provider|#12.0|training'                     
nid000019:54130:54690 [0] NCCL INFO NET/OFI Selected Provider is tcp;ofi_rxm
nid000020:73131:73750 [1] NCCL INFO NET/OFI Selected Provider is tcp;ofi_rxm
nid000020:73118:73752 [2] NCCL INFO NET/OFI Selected Provider is tcp;ofi_rxm
nid000020:73192:73754 [0] NCCL INFO NET/OFI Selected Provider is tcp;ofi_rxm
nid000020:73137:73748 [3] NCCL INFO NET/OFI Selected Provider is tcp;ofi_rxm
nid000019:54059:54686 [2] NCCL INFO NET/OFI Selected Provider is tcp;ofi_rxm
nid000019:54055:54687 [1] NCCL INFO NET/OFI Selected Provider is tcp;ofi_rxm
nid000019:54057:54692 [3] NCCL INFO NET/OFI Selected Provider is tcp;ofi_rxm
Step #1190      Loss: 0.015864
Step #1190      Loss: 0.029893
Step #1200      Loss: 0.000464
Step #1200      Loss: 0.017465
Step #1210      Loss: 0.050433
Step #1210      Loss: 0.050269
Step #1220      Loss: 0.015303
Step #1220      Loss: 0.014197
Step #1230      Loss: 0.002562
Step #1230      Loss: 0.007272
Step #1240      Loss: 0.066956
Step #1240      Loss: 0.011430
time to complete training: 80.6224 seconds
```
When running the OSU MPI bandwidth benchmark, utilizing `/host` libfabric, the performance numbers are quite respectable 
```
# env CUDA_VISIBLE_DEVICES=0,1,2,3 srun --mpi=pmi2 -n 2 --ntasks-per-node=1 -p grizzly singularity run --nv --bind /scratch2,/usr:/host/usr /scratch2/karlon/detAI-cuda-11.3-pytorch-1.10-tf-2.8-gpu-nccl-0.19.4.sif /scratch2/crickett/osu-micro-benchmarks-5.6.2/mpi/one-sided/osu_get_bw
srun: job 573423 queued and waiting for resources
srun: job 573423 has been allocated resources
# OSU MPI_Get Bandwidth Test v5.6.2
# Window creation: MPI_Win_allocate
# Synchronization: MPI_Win_flush
# Size      Bandwidth (MB/s)
1                       1.08
2                       2.15
4                       4.44
8                       8.86
16                     17.54
32                     35.08
64                     70.09
128                   140.23
256                   266.31
512                   514.50
1024                 1004.04
2048                 2004.23
4096                 3901.57
8192                 7288.89
16384               11934.33
32768               15784.98
65536               18843.25
131072              20809.72
262144              22361.31
524288              23258.59
1048576             23526.31
2097152             23715.48
4194304             23813.20
```

when running the same benchmark falling back to the TCP provider, the performance is quite lacking
```
# env NCCL_DEBUG=INFO CUDA_VISIBLE_DEVICES=0,1,2,3 srun --mpi=pmi2 -n 2 --ntasks-per-node=1 -p grizzly singularity run --nv --bind /scratc
h2 /scratch2/karlon/detAI-cuda-11.3-pytorch-1.10-tf-2.8-gpu-nccl-0.19.4.sif /scratch2/crickett/osu-micro-benchmarks-5.6.2/mpi/one-sided/osu_get_bw
INFO: /host not present, check srun for --bind /usr:/host/usr
INFO: /host not present, check srun for --bind /usr:/host/usr
# OSU MPI_Get Bandwidth Test v5.6.2
# Window creation: MPI_Win_allocate
# Synchronization: MPI_Win_flush
# Size      Bandwidth (MB/s)
1                       0.22
2                       0.43
4                       0.98
8                       1.89
16                      4.00
32                      7.35
64                     15.01
128                    30.33
256                    53.96
512                    87.60
1024                  148.25
2048                  224.42
4096                  256.16
8192                  278.84
16384                 299.78
32768                 303.26
65536                2913.55
131072               2988.84
262144               3598.28
524288               4101.77
1048576              4583.33
2097152              4855.23
4194304              5088.45
```

### CODE REVIEW

**Comment by stoksc** [comment]
@cla-bot check

File: dockerfile_scripts/scrape_libs.sh [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+host_dir="/host"
+if [ -d "$host_dir" ]; then
+    libfabric=`find /host -name libfabric.so 2>/dev/null`
+    libfabric_dir="$(dirname "$libfabric")"
+    if [[ ! -z "$libfabric" ]] ; then
+        # Need libfabric to be first in the LD_LIBRARY_PATH to
+        # override what is in the container. However, we likely
+        # need/want the libs that it is dependent upon to show up
+        # after the container ones so we append the rest of the
+        # host libs.  tmp_dir="`pwd`/tmp"
```
Comment (mpkouznetsov) [inline dockerfile_scripts/scrape_libs.sh:13]: Should the trailing `tmp_dir=...` here be removed from the comment?

File: Dockerfile-base-gpu [hunk 2]
```diff
@@ -52,19 +52,36 @@ ARG PYTHON_VERSION
...
+ARG OFI_LIB_DIR=${OFI_INSTALL_DIR}/lib:${OFI_INSTALL_DIR}/lib64
+ARG OFI_PATH_DIR=${OFI_INSTALL_DIR}/bin
 ARG OMPI_LIB_DIR=${OMPI_INSTALL_DIR}/lib
 ARG OMPI_PATH_DIR=${OMPI_INSTALL_DIR}/bin
-ENV LD_LIBRARY_PATH $LD_LIBRARY_PATH:${WITH_MPI:+$VERBS_LIB_DIR:$UCX_LIB_DIR:$OMPI_LIB_DIR}
-ENV PATH $PATH:${WITH_MPI:+$UCX_PATH_DIR:$OMPI_PATH_DIR}
+ARG AWS_LIB_DIR=${AWS_PLUGIN_INSTALL_DIR}/lib
+
+#USING OFI
+#ENV LD_LIBRARY_PATH=${WITH_OFI:+$LD_LIBRARY_PATH:${WITH_MPI:+$VERBS_LIB_DIR:$OFI_LIB_DIR:$OMPI_LIB_DIR:$AWS_LIB_DIR}}
```
Comment (mpkouznetsov) [inline Dockerfile-base-gpu:74]: Did you forget to remove the commented line?

**Comment by mpkouznetsov** [comment]
Hmm, the build fails. PATH seems to be messed up, and `python` executable is not found.

**Pull request closed without merging**